dpnp/dpnp_utils/dpnp_utils_pad.py

```diff
@@ -0,0 +1,784 @@
+
+    Parameters
+    ----------
+    sl : slice
+        The slice for the given dimension.
+    axis : int
+        The axis to which `sl` is applied. All other dimensions are left
+        "unsliced".
+
+    Returns
+    -------
+    sl : tuple of slices
+        A tuple with slices matching `shape` in length.
+
+    Examples
+    --------
+    >>> import dpnp as np
+    >>> import dpnp.dpnp_utils.dpnp_utils_pad as np_pad
+    >>> np_pad._slice_at_axis(slice(None, 3, -1), 1)
+    (slice(None, None, None), slice(None, 3, -1), Ellipsis)
+
+    """
+    return (slice(None),) * axis + (sl,) + (Ellipsis,)
+
+
+def _view_roi(array, original_area_slice, axis):
+    """
+    Copied from numpy/lib/_arraypad_impl.py
+
+    Gets a view of the current region of interest during iterative padding.
+
+    When padding multiple dimensions iteratively corner values are
+    unnecessarily overwritten multiple times. This function reduces the
+    working area for the first dimensions so that corners are excluded.
+
+    Parameters
+    ----------
+    array : dpnp.ndarray
+        The array with the region of interest.
+    original_area_slice : tuple of slices
+        Denotes the area with original values of the unpadded array.
+    axis : int
+        The currently padded dimension assuming that `axis` is padded before
+        `axis` + 1.
+
+    Returns
+    -------
+    roi : ndarray
+        The region of interest of the original `array`.
+
+    """
+
+    axis += 1
+    sl = (slice(None),) * axis + original_area_slice[axis:]
+    return array[sl]
+
+
+def dpnp_pad(array, pad_width, mode="constant", **kwargs):
+    """Pad an array."""
+
+    if isinstance(pad_width, int):
+        if pad_width < 0:
+            raise ValueError("index can't contain negative values")
+        pad_width = ((pad_width, pad_width),) * array.ndim
+    else:
+        if dpnp.is_supported_array_type(pad_width):
+            pad_width = dpnp.asnumpy(pad_width)
+        else:
+            pad_width = numpy.asarray(pad_width)
+
+        if not pad_width.dtype.kind == "i":
+            raise TypeError("`pad_width` must be of integral type.")
+
+        # Broadcast to shape (array.ndim, 2)
+        pad_width = _as_pairs(pad_width, array.ndim, as_index=True)
+
+    if callable(mode):
+        function = mode
+        # Create a new zero padded array
+        padded, _ = _pad_simple(array, pad_width, fill_value=0)
+        # And apply along each axis
+
+        for axis in range(padded.ndim):
+            # Iterate using ndindex as in apply_along_axis, but assuming that
+            # function operates inplace on the padded array.
+
+            # view with the iteration axis at the end
+            view = dpnp.moveaxis(padded, axis, -1)
+
+            # compute indices for the iteration axes, and append a trailing
+            # ellipsis to prevent 0d arrays decaying to scalars
+            inds = numpy.ndindex(view.shape[:-1])
```

Comment:
Could you please add a TODO comment to replace with `dpnp.ndindex` later (and create a task to backlog if missing)?